### DISCUSSION TITLE
Rectangular domain with circular cutout

### DISCUSSION THREAD

**Discussion by lindsayad** [idaholab/moose#21411] [Q&A Meshing] [answered]
Hi all, I'm looking to generate a rectangular domain with a circular cutout for vortex shedding analysis. What's the best way to do this using (existing) mesh generators? (@albertjp7) Maybe @GiudGiud has an idea?

**Comment by GiudGiud** [comment, marked as answer]
Use this
https://mooseframework.inl.gov/source/meshgenerators/ConcentricCircleMeshGenerator.html
then delete the circle(s) block(s)

Guillaume

**Reply by YaqiWang** [reply]
Or maybe this (https://github.com/idaholab/moose/pull/13696) if the concentric circle mesh is not ideal?

**Reply by lindsayad** [reply]
Maybe I need to add a `blocks` parameter? Otherwise I do not see how to directly control the block name/ID of the created circle and consequently know what block to delete

**Reply by GiudGiud** [reply]
the blocks have different ids anyway

but yes a blocks parameter would be welcome!

**Reply by lindsayad** [reply]
Hmm, I'm not sure how I'm gonna make this work. I want to make a domain for simulating [vortex shedding](https://en.wikipedia.org/wiki/Vortex_shedding) for which we want a long region after the circle such that outlet boundary condition does not affect the flow around the sphere. I'm not sure how I'm going to be able to stitch such an exit region (which I was thinking I would create just with a `GeneratedMeshGenerator`) with the (square - circle) region that the `ConcentricCircleMeshGenerator` gives me. Hard for me to see those nodes getting lined up

**Reply by lindsayad** [reply]
(this taken from our concentric circle mesh generator test)

![image](https://user-images.githubusercontent.com/10248304/176325955-246e697b-dbf4-4ccb-8318-0029033c6c10.png)

**Reply by lindsayad** [reply]
@grmnptr would you use gmsh for this mesh or would you suggest another open source tool?

**Reply by lindsayad** [reply]
Yea @albertjp7, I think we'll just go with gmsh on this one (probably don't need this long of an exit length though)

![image](https://user-images.githubusercontent.com/10248304/176329753-f55797b7-02c6-44ed-ace2-a6dd30b6cad0.png)

**Reply by lindsayad** [reply]
Might need to think about boundary layers around the circle (and maybe along the walls)

**Reply by GiudGiud** [reply]
the mesh is regular on the sides on the square, you should be able to stitch it alright.

You can also create boundary layers naturally by having an extra ring

**Reply by grmnptr** [reply]
I would use Salome, but gmsh is fine. You can definitely fit a structured mesh on this. Add a square around the circle and connect the corners to the circle that way you will end up with four quadrants which can be refined close to the circle.

**Reply by AlbertJP7** [reply]
How would I implement the generated mesh with the long exit length?

**Reply by lindsayad** [reply]
> the mesh is regular on the sides on the square, you should be able to stitch it alright.

I don't think the node spacing is uniform. The node positions are sine and cosine functions

**Reply by GiudGiud** [reply]
this works fine to me

```
[Mesh]
  [./ccmg]
    type = ConcentricCircleMeshGenerator
    num_sectors = 6
    radii = '0.2546 0.3368 0.3600 0.3818 0.3923 0.4025 0.4110 0.4750'
    rings = '5 3 2 1 1 1 1 3 5'
    has_outer_square = on
    pitch = 1
    #portion = left_half
    preserve_volumes = off
    smoothing_max_it = 3
  []
  [left]
    type = CartesianMeshGenerator
    dim = 2
    dx = '10'
    dy = '1'
    ix = '10'
    iy = '18'
  []
  [move_it]
    type = TransformGenerator
    input = left
    transform = translate
    vector_value = '-10.5 -0.5 0'
  []
  [stitch]
    type = StitchedMeshGenerator
    inputs = 'ccmg move_it'
    stitch_boundaries_pairs = 'right left'
  []
[]
```

**Reply by lindsayad** [reply]
How did you determine `iy` here? Did you figure it out from the logic in `ConcentricCircleMeshGenerator`? If so, it might be good to add somewhere in the documentation what the formula is (I would be more than willing to add it). My dumb brain would have to sit down and run through the for loop logic in there ... but if you already know it, then my dumb brain doesn't have to :smile:

**Reply by GiudGiud** [reply]
I counted :)

**Reply by GiudGiud** [reply]
looks like my brain is dumber than yours haha

it does look like we can document this object better

**Reply by doquang** [reply]
@lindsayad , @GiudGiud - Do you have a similar thing as ConcentricCircleMeshGenerator for 3D?

**Reply by GiudGiud** [reply]
like for a sphere?

**Reply by GiudGiud** [reply]
or a cylinder

**Reply by doquang** [reply]
For a cylinder. Is it posible?

**Reply by GiudGiud** [reply]
yes just use an extruder after
https://mooseframework.inl.gov/source/meshgenerators/MeshExtruderGenerator.html

**Reply by doquang** [reply]
Any suggestions for meshing a cavity inside a box as
<img width="580" alt="image" src="https://user-images.githubusercontent.com/3186923/224073444-c3392437-9d3d-4c0e-b85f-2d27d01f5d6f.png">

**Reply by GiudGiud** [reply]
Is the cavity spherical?

**Reply by doquang** [reply]
I would like to start the cavity as a spherical, but for a long run, we would like to define the cavity from three polygons:
![image](https://user-images.githubusercontent.com/3186923/224082619-20c3483a-301d-49e3-ad49-8f3125afc753.png)

**Reply by GiudGiud** [reply]
You can use this
https://mooseframework.inl.gov/source/meshgenerators/SphereMeshGenerator.html
then use a blockdeletionGenerator to remove the sphere


You could deform the sphere using a ParsedNodeTransformGenerator

**Reply by doquang** [reply]
@GiudGiud - I haven't found parameter  `has_outer_square = on` on `SphereMeshGenerator` as in `ConcentricCircleMeshGenerator`. How can I generate the mesh between the cavity surface to the surfaces of the box?

**Reply by GiudGiud** [reply]
oh then it wont work.
You ll have to use an external mesher
